### ISSUE TITLE
Support for lifted nn.switch module

### ISSUE THREAD

**Issue by marcociccone** [google/flax#2200]
Hi,  I think it would be useful to add a lifted version of `jax.lax.switch` to flax linen as done for other libraries such as haiku.
As a use case consider having a list of functions (in our case `nn` modules) and you want to apply them over an axis of your input. https://github.com/google/jax/issues/673#issuecomment-894955037

See the [original](https://colab.research.google.com/gist/tomhennigan/86fe7b7d46a930b58a5d8d76a75d75ea/fix-for-300.ipynb) haiku example from https://github.com/deepmind/dm-haiku/issues/300#issuecomment-1028195090 that I readapted to how I expect it to work in flax.

````python
class MultiDense(nn.Module):
    features: nn.Module
    num_classes: int
    num_heads: int = 1
    dtype: Any = jnp.float32 

    @nn.compact
    def __call__(self,
                 x,
                 train: bool = True,
                 heads_idx: List = None):
        
        # first time calling for init shoudl pass all indexes 
        self.heads = [nn.Dense(num_classes, dtype=self.dtype) for range(num_heads)]
        vmap_fns = nn.vmap(lambda i, x: nn.switch(i, self.heads, x))
        x = vmap_fns(heads_idx, x)
        return x

**Comment by cgarciae** [comment]
Implemented in #2205.

**Comment by cgarciae** [comment]
See example how to implement multi-head architecture here: https://flax.readthedocs.io/en/latest/_autosummary/flax.linen.switch.html?highlight=switch#flax.linen.switch

**Comment by jheek** [comment]
Btw do keep in mind that if you vmap over the switch index than jax will rewrite the switch into a select meaning it will execute *all the branches for each input* so in this example the conditional doesn't actually avoid any computation

**Comment by jheek** [comment]
You could use an nn.scan together with an nn.cond but that only works efficiently if the heads are large enough and make efficient use of flops if they aren't batched together
